File: gren/graphs/reducers/audit-report.py
# Deterministic markdown audit report from inventory, verified issues, fix plans and dismissed issues.
# input: { repo, inventory, route, issues (verified, prioritised), fixes (survivors), dismissed (killed), unverified_fixes (killed fix plans), stats }
import json


def reduce(input, args, ctx):
    inv = input.get("inventory") or {}
    issues = input.get("issues") if isinstance(input.get("issues"), list) else []
    fixes = input.get("fixes") if isinstance(input.get("fixes"), list) else []
    dismissed = input.get("dismissed") if isinstance(input.get("dismissed"), list) else []
    bad_fixes = input.get("unverified_fixes") if isinstance(input.get("unverified_fixes"), list) else []
    L: list[str] = [f"# Codebase audit: {input.get('repo') or inv.get('root') or 'repository'}", ""]
    mods = inv.get("modules")
    L += [f"> {inv.get('file_count', '?')} files · {inv.get('total_lines', '?')} lines · {len(mods) if isinstance(mods, list) else '?'} modules · route: {input.get('route') or '?'} · {len(issues)} verified issues · {len(dismissed)} dismissed by the verifier · {len(fixes)} fix plans survived review", ""]
    if isinstance(input.get("stats"), dict):
        L += ["> " + " · ".join(f"{k}: {json.dumps(v) if isinstance(v, (dict, list)) else v}" for k, v in input["stats"].items()), ""]
    L += ["## Verified issues (prioritised)", ""]
    if not issues:
        L += ["_No issue survived verification._", ""]
    for it in issues:
        L += [f"### {it.get('rank', '')}. [{str(it.get('severity')).upper()}] {it.get('title')}", "", f"**Where:** `{it.get('file')}{':' + str(it['line']) if it.get('line') else ''}` · **Category:** {it.get('category')}", "", str(it.get("description") or "").strip(), ""]
        if it.get("evidence"):
            L += ["```", str(it["evidence"]).strip(), "```", ""]
        if it.get("suggested_fix"):
            L += [f"**Suggested fix:** {it['suggested_fix']}", ""]
    L += ["## Fix plans (survived adversarial review)", ""]
    if not fixes:
        L += ["_No fix plan survived review._", ""]
    for f in fixes:
        L += [f"### {f.get('issue_ref') or f.get('title') or 'fix'}", "", str(f.get("plan") or "").strip(), ""]
        if f.get("patch"):
            L += ["```diff", str(f["patch"]).strip(), "```", ""]
        if f.get("risk"):
            L += [f"**Risk:** {f['risk']}", ""]
        if isinstance(f.get("tests_to_add"), list) and f["tests_to_add"]:
            L += ["**Tests to add:**"] + [f"- {t}" for t in f["tests_to_add"]] + [""]
    if bad_fixes:
        L += ["## Fix plans rejected by review", ""]
        for k in bad_fixes:
            item = k.get("item") or {}
            L.append(f"- {item.get('issue_ref') or item.get('title') or '#' + str(k.get('index'))}: {'; '.join(k.get('reasons') or [])}")
        L.append("")
    if dismissed:
        L += ["## Reported issues dismissed by the verifier", ""]
        for k in dismissed:
            item = k.get("item") or {}
            L.append(f"- `{item.get('file') or '?'}` {item.get('title') or ''}: {'; '.join(k.get('reasons') or [])}")
        L.append("")
    ctx.log(f"audit report: {len(issues)} issues, {len(fixes)} fixes, {len(dismissed)} dismissed")
    return {"markdown": "\n".join(L), "issues": len(issues), "fixes": len(fixes), "dismissed": len(dismissed), "_stats": {"in": len(issues) + len(dismissed), "out": len(issues)}}

File: gren/v1/graphs/reducers/audit-report.js
// Deterministic markdown audit report from inventory, verified issues, fix plans and dismissed issues.
// input: { repo, inventory, route, issues (verified, prioritised), fixes (survivors), dismissed (killed), unverified_fixes (killed fix plans), stats }
export default async function auditReport(input, args, ctx) {
  const L = [];
  const inv = input.inventory ?? {};
  const issues = Array.isArray(input.issues) ? input.issues : [];
  const fixes = Array.isArray(input.fixes) ? input.fixes : [];
  const dismissed = Array.isArray(input.dismissed) ? input.dismissed : [];
  const badFixes = Array.isArray(input.unverified_fixes) ? input.unverified_fixes : [];
  L.push(`# Codebase audit: ${input.repo ?? inv.root ?? "repository"}`, "");
  L.push(`> ${inv.file_count ?? "?"} files · ${inv.total_lines ?? "?"} lines · ${Array.isArray(inv.modules) ? inv.modules.length : "?"} modules · route: ${input.route ?? "?"} · ${issues.length} verified issues · ${dismissed.length} dismissed by the verifier · ${fixes.length} fix plans survived review`, "");
  if (input.stats) L.push(`> ${Object.entries(input.stats).map(([k, v]) => `${k}: ${typeof v === "object" ? JSON.stringify(v) : v}`).join(" · ")}`, "");
  L.push("## Verified issues (prioritised)", "");
  if (!issues.length) L.push("_No issue survived verification._", "");
  for (const it of issues) {
    L.push(`### ${it.rank ?? ""}. [${String(it.severity).toUpperCase()}] ${it.title}`, "", `**Where:** \`${it.file}${it.line ? `:${it.line}` : ""}\` · **Category:** ${it.category}`, "", String(it.description ?? "").trim(), "");
    if (it.evidence) L.push("```", String(it.evidence).trim(), "```", "");
    if (it.suggested_fix) L.push(`**Suggested fix:** ${it.suggested_fix}`, "");
  }
  L.push("## Fix plans (survived adversarial review)", "");
  if (!fixes.length) L.push("_No fix plan survived review._", "");
  for (const f of fixes) {
    L.push(`### ${f.issue_ref ?? f.title ?? "fix"}`, "", String(f.plan ?? "").trim(), "");
    if (f.patch) L.push("```diff", String(f.patch).trim(), "```", "");
    if (f.risk) L.push(`**Risk:** ${f.risk}`, "");
    if (Array.isArray(f.tests_to_add) && f.tests_to_add.length) L.push("**Tests to add:**", ...f.tests_to_add.map((t) => `- ${t}`), "");
  }
  if (badFixes.length) {
    L.push("## Fix plans rejected by review", "");
    for (const k of badFixes) L.push(`- ${k.item?.issue_ref ?? k.item?.title ?? `#${k.index}`}: ${(k.reasons ?? []).join("; ")}`);
    L.push("");
  }
  if (dismissed.length) {
    L.push("## Reported issues dismissed by the verifier", "");
    for (const k of dismissed) L.push(`- \`${k.item?.file ?? "?"}\` ${k.item?.title ?? ""}: ${(k.reasons ?? []).join("; ")}`);
    L.push("");
  }
  ctx.log(`audit report: ${issues.length} issues, ${fixes.length} fixes, ${dismissed.length} dismissed`);
  return { markdown: L.join("\n"), issues: issues.length, fixes: fixes.length, dismissed: dismissed.length, _stats: { in: issues.length + dismissed.length, out: issues.length } };
}
